File: harvester/src/main/java/harvester/ui/HarvesterTableModel.java
package harvester.ui;

import harvester.RelatedKeyword;

import javax.swing.SwingUtilities;
import javax.swing.table.AbstractTableModel;

public class HarvesterTableModel extends AbstractTableModel {

    private static final long serialVersionUID = -5057681513894708028L;
    private RelatedKeyword[] keywords = new RelatedKeyword[0];

    @Override
    public String getColumnName(int column) {
	switch (column) {
	case 0:
	    return "Word";

	case 1:
	    return "Relevance";
	}
	throw new IllegalArgumentException();
    }

    @Override
    public int getRowCount() {
	return keywords.length;
    }

    @Override
    public int getColumnCount() {
	return 2;
    }

    @Override
    public Object getValueAt(int rowIndex, int columnIndex) {
	RelatedKeyword rk = keywords[rowIndex];
	switch (columnIndex) {
	case 0:
	    return rk.getWord();
	case 1:
	    return rk.getRelevance();
	}
	throw new IllegalArgumentException();
    }

    synchronized void setData(RelatedKeyword[] keywords) {
	this.keywords = keywords;
	SwingUtilities.invokeLater(new Runnable() {
	    @Override
	    public void run() {
		fireTableDataChanged();
	    }
	});
    }
}

File: harvester-scala/src/main/scala/harvester/HarvesterTableModel.scala
package harvester;

import javax.swing.table.AbstractTableModel
import javax.swing.SwingUtilities

import scala.collection.JavaConversions._
object HarvesterTableModel{}
class HarvesterTableModel extends AbstractTableModel {
  var keywords = Array.empty[RelatedKeyword]

  override def getColumnName(column: Int) = {
    column match {
      case 0 => "Word"
      case 1 => "Relevance"
    }
  }

  def getRowCount = keywords.size

  def getColumnCount = 2

  
  override def getValueAt(rowIndex: Int, columnIndex: Int) = {
    columnIndex match {
      case 0 => keywords(rowIndex).word
      case 1 => keywords(rowIndex).relevance.asInstanceOf[Object]
    }
  }

  def setData(keywords: Array[RelatedKeyword]) {
    this.keywords = keywords;
    MySwingUtilities.invokeLater(fireTableDataChanged)
  }
}

object MySwingUtilities {
  def invokeLater[X](exp: => X) {
    import javax.swing.SwingUtilities

    SwingUtilities invokeLater (new Runnable() {
      def run = exp
    })
  }
}
